index/mem/index.go

```diff
@@ -0,0 +1,302 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"crypto/md5"
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index"
+
+	"github.com/RoaringBitmap/roaring"
+	"github.com/uber-go/atomic"
+)
+
+const initialSize = 1024 * 1024
+
+// TODO(prateek): investigate impact of native heap
+type memIndex struct {
+	opts   Options
+	hashFn doc.HashFn
+
+	docIDGen *atomic.Uint32
+
+	// hash(doc.ID)  -> internal docID
+	idsLock sync.RWMutex
+	idsMap  map[doc.Hash]index.DocID
```

Comment:
Is it perhaps worthwhile to make this a two-level map? The first level being just like it is now, hash(doc.ID) to a struct with the internal DocID and the doc.ID, the second level then being doc.ID to internal DocID. This way we can guarantee that even if two id's have the same hash we're safe. Also, we can probably reduce the size of the hash to 64 bits since we have a failsafe.

Reply:
tl;dr ~ yeah we need to, will put down a TODO. 

I'm using the map for two things:
1) as a cheap "don't write cache",
2) for cheaper updates because we were able to do a direct lookup by doc hash. 

For (1) I'd actually like this to be a LRUCache so we can treat it as a reverse bloom-filter. I'll put down a TODO for this too. 

For (2) I'm getting more inclined to treat all updates as a straight insert. Even the in-memory ones. Again handling the dedupe read side. 

p.s. (2) made me realise we're going to need some kind of `Optimize()` call to cleanup dead/redundant doc references. Will add that to the `Writable` interface.

Reply:
Moved this map to the struct backing the `Index` interface, so it can be shared amongst all segments. Left a todo explaining this discussion.